src/handlers/next/process-protected-handler.ts

```diff
@@ -0,0 +1,106 @@
+import { NextApiRequest } from "next";
+
+import { APL } from "../../APL";
+import { AuthData } from "../../APL/apl";
+import { createDebug } from "../../debug";
+import { getAppId } from "../../get-app-id";
+import { getBaseUrl, getSaleorHeaders } from "../../headers";
+import { verifyJWT } from "../../verify-jwt";
+
+const debug = createDebug("processProtectedHandler");
+
+export type SaleorProtectedHandlerError =
+  | "OTHER"
+  | "MISSING_HOST_HEADER"
+  | "MISSING_DOMAIN_HEADER"
+  | "MISSING_AUTHORIZATION_BEARER_HEADER"
+  | "NOT_REGISTERED"
+  | "JWT_VERIFICATION_FAILED"
+  | "NO_APP_ID";
+
+export class ProtectedHandlerError extends Error {
+  errorType: SaleorProtectedHandlerError = "OTHER";
+
+  constructor(message: string, errorType: SaleorProtectedHandlerError) {
+    super(message);
+    if (errorType) {
+      this.errorType = errorType;
+    }
+    Object.setPrototypeOf(this, ProtectedHandlerError.prototype);
+  }
+}
+
+export type ProtectedHandlerContext = {
+  baseUrl: string;
+  authData: AuthData;
+};
+
+interface ProcessSaleorProtectedHandlerArgs {
+  req: NextApiRequest;
+  apl: APL;
+}
+
+type ProcessAsyncSaleorProtectedHandler = (
+  props: ProcessSaleorProtectedHandlerArgs
+) => Promise<ProtectedHandlerContext>;
+
+/**
+ * Perform security checks on given request and return ProtectedHandlerContext object.
+ * In case of validation issues, instance of the ProtectedHandlerError will be thrown.
+ */
+export const processSaleorProtectedHandler: ProcessAsyncSaleorProtectedHandler = async ({
+  req,
+  apl,
+}: ProcessSaleorProtectedHandlerArgs): Promise<ProtectedHandlerContext> => {
+  debug("Request processing started");
+  const { domain, authorizationBearer: token } = getSaleorHeaders(req.headers);
+
+  const baseUrl = getBaseUrl(req.headers);
+  if (!baseUrl) {
+    debug("Missing host header");
+    throw new ProtectedHandlerError("Missing host header", "MISSING_HOST_HEADER");
+  }
+
+  if (!domain) {
+    debug("Missing saleor-domain header");
+    throw new ProtectedHandlerError("Missing saleor-domain header", "MISSING_DOMAIN_HEADER");
+  }
+
+  if (!token) {
+    debug("Missing authorization-bearer header");
+    throw new ProtectedHandlerError(
+      "Missing authorization-bearer header",
+      "MISSING_AUTHORIZATION_BEARER_HEADER"
+    );
+  }
```

Comment:
Not scope of this PR but maybe we should write some common assertions library, that will share this logic, eg

```
assert.authHeader(req); // void or throw
```

We will have unified errors which are technically the same, we can add context of which handler in extra argument

Reply:
Love the idea. Added issue #125